File: problems/algorithms/0509/C/solution_recursive_memoization.c
int calculated[31] = {0};

int fib(int n)
{
	if (n < 2)
		return n;
	if (calculated[n] != 0)
		return calculated[n];
	return calculated[n] = fib(n - 1) + fib(n - 2);
}

File: problems/algorithms/0509/Rust/solution_recursive_memoization.rs
impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n < 2 {
            return n;
        }
        let i = n as usize;
        let mut already_calculated = vec![0; i + 1];
        Solution::fib_helper(i - 1, &mut already_calculated) + Solution::fib_helper(i - 2, &mut already_calculated)
    }
    
    fn fib_helper(n: usize, calculated: &mut Vec<i32>) -> i32 {
        if n < 2 {
            return n as i32;
        }
        if calculated[n] != 0 {
            return calculated[n];
        }
        calculated[n] = Solution::fib_helper(n - 1, calculated) + Solution::fib_helper(n - 2, calculated);
        calculated[n]
    }
}